pairing/bn254/suite.go

```diff
@@ -0,0 +1,209 @@
+// package bn254 implements a particular bilinear group.
+//
+// Bilinear groups are the basis of many of the new cryptographic protocols that
+// have been proposed over the past decade. They consist of a triplet of groups
+// (G₁, G₂ and GT) such that there exists a function e(g₁ˣ,g₂ʸ)=gTˣʸ (where gₓ
+// is a generator of the respective group). That function is called a pairing
+// function.
+//
+// This package specifically implements the Optimal Ate pairing over a 256-bit
+// Barreto-Naehrig curve as described in
+// http://cryptojedi.org/papers/dclxvi-20100714.pdf. Its output is compatible
+// with the implementation described in that paper.
+//
+// This package previously claimed to operate at a 128-bit security level.
+// However, recent improvements in attacks mean that is no longer true. See
+// https://moderncrypto.org/mail-archive/curves/2016/000740.html.
+package bn254
+
+import (
+	"crypto/cipher"
+	"hash"
+	"io"
+	"reflect"
+
+	"github.com/drand/kyber"
+	"github.com/drand/kyber/util/random"
+	"github.com/drand/kyber/xof/blake2xb"
+	"go.dedis.ch/fixbuf"
+	"golang.org/x/crypto/sha3"
+)
+
+// Suite implements the pairing.Suite interface for the BN254 bilinear pairing.
+type Suite struct {
+	*commonSuite
+	g1 *groupG1
+	g2 *groupG2
+	gt *groupGT
+}
+
+var DEFAULT_DOMAIN_G1 = []byte("BN254G1_XMD:KECCAK-256_SSWU_RO_")
+var DEFAULT_DOMAIN_G2 = []byte("BN254G2_XMD:KECCAK-256_SSWU_RO_")
+
+// NewSuite generates and returns a new BN254 pairing suite.
+func NewSuite() *Suite {
+	s := &Suite{commonSuite: &commonSuite{}}
+	s.g1 = &groupG1{
+		commonSuite: s.commonSuite,
+		dst:         DEFAULT_DOMAIN_G1,
+	}
+	s.g2 = &groupG2{
+		commonSuite: s.commonSuite,
+		dst:         DEFAULT_DOMAIN_G2,
+	}
+	s.gt = &groupGT{commonSuite: s.commonSuite}
+	return s
+}
+
+// NewSuiteG1 returns a G1 suite.
+func NewSuiteG1() *Suite {
+	s := NewSuite()
+	s.commonSuite.Group = &groupG1{commonSuite: &commonSuite{}}
+	return s
+}
+
+// NewSuiteG2 returns a G2 suite.
+func NewSuiteG2() *Suite {
+	s := NewSuite()
+	s.commonSuite.Group = &groupG2{commonSuite: &commonSuite{}}
+	return s
+}
+
+// NewSuiteGT returns a GT suite.
+func NewSuiteGT() *Suite {
+	s := NewSuite()
+	s.commonSuite.Group = &groupGT{commonSuite: &commonSuite{}}
+	return s
+}
+
+// NewSuiteRand generates and returns a new BN254 suite seeded by the
+// given cipher stream.
+func NewSuiteRand(rand cipher.Stream) *Suite {
+	s := &Suite{commonSuite: &commonSuite{s: rand}}
+	s.g1 = &groupG1{
+		commonSuite: s.commonSuite,
+		dst:         DEFAULT_DOMAIN_G1,
+	}
+	s.g2 = &groupG2{
+		commonSuite: s.commonSuite,
+		dst:         DEFAULT_DOMAIN_G2,
+	}
+	s.gt = &groupGT{commonSuite: s.commonSuite}
+	return s
+}
+
+// Set G1 DST
+func (s *Suite) SetDomainG1(dst []byte) {
+	s.g1.dst = dst
+}
+
+// Set G2 DST
+func (s *Suite) SetDomainG2(dst []byte) {
+	s.g2.dst = dst
```

Comment:
I wonder if it wouldn't be safer to copy the byte content into `.dst` to avoid any risk of reference messing up the dst everywhere afterwards?

See:
https://play.golang.com/p/mAiEzzKzcaV

Reply:
Makes sense - using `copy` now. I also changed where we set the default DSTs so that the defaults are not mutable.